### PR TITLE
Auto assign numbers for Enum elements

### PR DESCRIPTION
**Pull request by awakeljw**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Auto assign numbers for Enum elements

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
@awakeljw though I'm not sure if we should allow mix syntax - there is some inconsistency:
```
CREATE TABLE auto_assgin_enum
(
    `x` enum('a' = 2, 'b')
)
ENGINE = MergeTree
ORDER BY x

Query id: af3d776f-1466-42b3-b9c8-aa1dd6a66042

Ok.

0 rows in set. Elapsed: 0.010 sec.

ClickHouse-ubuntu :) describe auto_assgin_enum

DESCRIBE TABLE auto_assgin_enum

Query id: 3010e516-ad9b-46ca-93ca-6f04292eebf2

┌─name─┬─type────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ x    │ Enum8('b' = 1, 'a' = 2) │              │                    │         │                  │                │
└──────┴─────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

1 rows in set. Elapsed: 0.005 sec.
```

and:
```
CREATE TABLE auto_assgin_enum
(
    `x` enum('a', 'b' = 1)
)
ENGINE = MergeTree
ORDER BY x

Query id: 25ea3eab-84cd-4f9c-bd61-e701e2687641


0 rows in set. Elapsed: 0.673 sec.

Received exception from server (version 22.4.1):
Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Duplicate values in enum: 'a' = 1 and '1'. (SYNTAX_ERROR)```

**Comment by yakov-olkhovskiy** [comment]
...even more confusing:
```
ClickHouse-ubuntu :) CREATE TABLE auto_assgin_enum (x enum('a'=3, 'b', 'c')) ENGINE=MergeTree() order by x;

CREATE TABLE auto_assgin_enum
(
    `x` enum('a' = 3, 'b', 'c')
)
ENGINE = MergeTree
ORDER BY x

Query id: da628b77-2f94-46a7-821a-24500c2f8c5d

Ok.

0 rows in set. Elapsed: 0.041 sec.

ClickHouse-ubuntu :) DROP TABLE IF EXISTS auto_assgin_enum;

DROP TABLE IF EXISTS auto_assgin_enum

Query id: 5325c560-f37e-4adb-ac2e-a059a09c5b27

Ok.

0 rows in set. Elapsed: 0.002 sec.

ClickHouse-ubuntu :) CREATE TABLE auto_assgin_enum (x enum('a'=3, 'b', 'c', 'd')) ENGINE=MergeTree() order by x;

CREATE TABLE auto_assgin_enum
(
    `x` enum('a' = 3, 'b', 'c', 'd')
)
ENGINE = MergeTree
ORDER BY x

Query id: a116f722-5625-4967-92ef-f77adfb5b63c


0 rows in set. Elapsed: 0.004 sec.

Received exception from server (version 22.4.1):
Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Duplicate values in enum: 'a' = 3 and '3'. (SYNTAX_ERROR)
```

**Comment by awakeljw** [comment]
> 



> ...even more confusing:
> 
> ```
> ClickHouse-ubuntu :) CREATE TABLE auto_assgin_enum (x enum('a'=3, 'b', 'c')) ENGINE=MergeTree() order by x;
> 
> CREATE TABLE auto_assgin_enum
> (
>     `x` enum('a' = 3, 'b', 'c')
> )
> ENGINE = MergeTree
> ORDER BY x
> 
> Query id: da628b77-2f94-46a7-821a-24500c2f8c5d
> 
> Ok.
> 
> 0 rows in set. Elapsed: 0.041 sec.
> 
> ClickHouse-ubuntu :) DROP TABLE IF EXISTS auto_assgin_enum;
> 
> DROP TABLE IF EXISTS auto_assgin_enum
> 
> Query id: 5325c560-f37e-4adb-ac2e-a059a09c5b27
> 
> Ok.
> 
> 0 rows in set. Elapsed: 0.002 sec.
> 
> ClickHouse-ubuntu :) CREATE TABLE auto_assgin_enum (x enum('a'=3, 'b', 'c', 'd')) ENGINE=MergeTree() order by x;
> 
> CREATE TABLE auto_assgin_enum
> (
>     `x` enum('a' = 3, 'b', 'c', 'd')
> )
> ENGINE = MergeTree
> ORDER BY x
> 
> Query id: a116f722-5625-4967-92ef-f77adfb5b63c
> 
> 
> 0 rows in set. Elapsed: 0.004 sec.
> 
> Received exception from server (version 22.4.1):
> Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Duplicate values in enum: 'a' = 3 and '3'. (SYNTAX_ERROR)
> ```

Thanks for your review. I have checked ENUM type in Mysql,it is
```
mysql> create table t1(a int, b ENUM('a', 'b'));
Query OK, 0 rows affected (0.17 sec)

mysql> create table t1(a int, b ENUM('a'=1, 'b'));
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=1, 'b'))' at line 1
mysql> create table t1(a int, b ENUM('a'=1, 'b'=2));
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '=1, 'b'=2))' at line 1
```
Maybe we should keep consistency, if Either "all Enum is String=Number" or "all Enum is String" is allowed, otherwise throw Exception.

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
@awakeljw everything looks good to me, you can merge now if you finished
thank you for your contribution

**Comment by awakeljw** [comment]
> @awakeljw everything looks good to me, you can merge now if you finished thank you for your contribution

I also wonder if Int16 is appropriate, because Enum16 is up to Int16, if number of value exceed 32767, it will change to -32768. Actually the ("name"=number) has 32768 ("1-32767" + "-32768"), "name" has 32768,(-32767~0), it will succeed.So i change to UInt64, it is safe now.  I have finish it now.
```
+    UInt64 literal_child_count = 0;
+    UInt64 func_child_count = 0;
```

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
From docs:
```
ClickHouse supports:
8-bit Enum. It can contain up to 256 values enumerated in the [-128, 127] range.
16-bit Enum. It can contain up to 65536 values enumerated in the [-32768, 32767] range.
...
In RAM, an Enum column is stored in the same way as Int8 or Int16 of the corresponding numerical values.
```
but this fix is ok, you can merge it after CI finished

File: src/DataTypes/DataTypeEnum.cpp [hunk 1]
```diff
@@ -191,6 +191,33 @@ static void checkASTStructure(const ASTPtr & child)
...
+        }
+        else
+        {
+            ++func_child_count;
+            assign_number_child.emplace_back(child);
+        }
+    }
+
+    if (func_child_count > 0 && literal_child_count > 0)
+        throw Exception("ALL Elements of Enum data type must be of form: 'name' = number or 'name', where name is string literal and number is an integer",
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeEnum.cpp:215]: Strange exception message, wrong syntax.

Why ALL is all caps?
Why Element with capital E?

File: tests/queries/0_stateless/00757_enum_defaults.sql [hunk 2]
```diff
@@ -1,2 +1,12 @@
 select os_name, count() from (SELECT CAST('iphone' AS Enum8('iphone' = 1, 'android' = 2)) AS os_name) group by os_name WITH TOTALS;
 select toNullable(os_name) AS os_name, count() from (SELECT CAST('iphone' AS Enum8('iphone' = 1, 'android' = 2)) AS os_name) group by os_name WITH TOTALS;
+
+DROP TABLE IF EXISTS auto_assgin_enum;
+DROP TABLE IF EXISTS auto_assgin_enum1;
+
+CREATE TABLE auto_assgin_enum (x enum('a', 'b')) ENGINE=MergeTree() order by x;
+CREATE TABLE auto_assgin_enum1 (x enum('a' = 1, 'b')) ENGINE=MergeTree() order by x; -- { serverError 223 }
+INSERT INTO auto_assgin_enum VALUES('a'), ('b');
+select * from auto_assgin_enum;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/00757_enum_defaults.sql:10]: The test does not check the assigned numbers.
You can use CAST for it.

File: src/DataTypes/DataTypeEnum.cpp (hunk 1 above)
Comment (yakov-olkhovskiy) [reply src/DataTypes/DataTypeEnum.cpp:215]: missed this one...

File: tests/queries/0_stateless/00757_enum_defaults.sql (hunk 2 above)
Comment (yakov-olkhovskiy) [reply tests/queries/0_stateless/00757_enum_defaults.sql:10]: should we check assigned numbers? it's straight forward increment from 1, and is it important? the only expected result - it's a set of unique numbers hidden behind names

Comment (yakov-olkhovskiy) [reply tests/queries/0_stateless/00757_enum_defaults.sql:10]: @alexey-milovidov BTW it's an interesting question - underlying enum type is signed, however usually enums are used with positive numbers - so effectively the available range is cut in half by users practice. Besides if we want to test numbers assignment above 8 bit range - it's not above 255 - it's above 127...

Comment (awakeljw) [reply tests/queries/0_stateless/00757_enum_defaults.sql:10]: I check for mysql, it's enum value increase from 1，so if we execute "select * from auto_assign_enum where x=1;" it should return 1
```
mysql> show create table t1;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                    |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `a` int DEFAULT NULL,
  `b` enum('a','b') DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select * from t1 where b = 1;
+------+------+
| a    | b    |
+------+------+
|    1 | a    |
+------+------+
1 row in set (0.00 sec)

mysql> select * from t1 where b = 2;
+------+------+
| a    | b    |
+------+------+
|    1 | b    |
+------+------+
1 row in set (0.00 sec)
```

File: src/DataTypes/DataTypeEnum.cpp (hunk 1 above)
Comment (awakeljw) [reply src/DataTypes/DataTypeEnum.cpp:215]: I will fix it, I just want to emphasize it....

File: tests/queries/0_stateless/00757_enum_defaults.sql (hunk 2 above)
Comment (awakeljw) [reply tests/queries/0_stateless/00757_enum_defaults.sql:10]: > 

I will add tests.

**Pull request merged**